### PR TITLE
With center=True in Encoder and Decoder

### PR DESCRIPTION
**Pull request by mpariente**
Trying to add `center=True` in `Encoder` and `Decoder` but I'm not fully satisfied for the `Decoder`. 

For the `Encoder`, I think it's fine. It just add two frames, that will be removed afterwards. 
For the `Decoder`, I'll comment inline.

Note: the base of the PR is `torch_compat`

### CODE REVIEW

File: asteroid_filterbanks/enc_dec.py [hunk 1]
```diff
@@ -130,15 +131,26 @@ class Encoder(_EncDec):
...
         self.as_conv1d = as_conv1d
+        self.center = center
+        if padding and center:
+            warnings.warn(f"Found `center=True` and `padding={padding}`. Padding takes precedence.")
+            self.padding = padding
+        elif center:
+            self.padding = [self.filterbank.kernel_size // 2, self.filterbank.kernel_size // 2]
+        else:
+            self.padding = padding
+        self.pad_mode = pad_mode
```
Comment (mpariente) [inline asteroid_filterbanks/enc_dec.py:152]: This is not very pretty but I don't know what else would make sense.

File: asteroid_filterbanks/enc_dec.py [hunk 2]
```diff
@@ -175,23 +187,36 @@ def forward(self, waveform):
             filters=filters,
             stride=self.stride,
             padding=self.padding,
+            pad_mode=self.pad_mode,
             as_conv1d=self.as_conv1d,
         )
         return self.filterbank.post_analysis(spec)
 
 
+@script_if_tracing
```
Comment (mpariente) [inline asteroid_filterbanks/enc_dec.py:196]: I'd have to make sure this is scriptable.

File: asteroid_filterbanks/enc_dec.py [hunk 3]
```diff
@@ -130,15 +131,26 @@ class Encoder(_EncDec):
...
             If False, will output a tensor of shape (batch, 1, freq, conv_time).
-        padding (int): Zero-padding added to both sides of the input.
+        padding (Union[int, List[int]): Zero-padding added to both sides of the input.
+        pad_mode (str): See torch.nn.functional.pad.
 
     """
 
-    def __init__(self, filterbank, is_pinv=False, as_conv1d=True, padding=0):
+    def __init__(
+        self, filterbank, is_pinv=False, as_conv1d=True, center=False, padding=0, pad_mode="reflect"
```
Comment (mpariente) [inline asteroid_filterbanks/enc_dec.py:140]: By the way, pad_mode=reflect was not the default before, it was constant. But if we want to match torch.stft, they use reflect. To be discussed.

File: asteroid_filterbanks/enc_dec.py [hunk 4]
```diff
@@ -283,7 +320,10 @@ def forward(self, spec, length: Optional[int] = None) -> torch.Tensor:
             padding=self.padding,
             output_padding=self.output_padding,
         )
+
         wav = self.filterbank.post_synthesis(wav)
+        start = self.filterbank.kernel_size // 2 if self.center else 0
+        wav = wav[..., start:]
```
Comment (mpariente) [inline asteroid_filterbanks/enc_dec.py:326]: If we remove the `start` before, the w**2 OLA computation is wrong. So, that's not an option. 
But if we do it after, then we had to divide by w**2 OLA on the sides as well, where it's most likely zero (hence the warning instead of the error). 
Not super satisfied about this, because we'll raise warnings about things that might not be problematic (if we trim the part with zeros after the division, which is the most common).

**Comment by mpariente** [comment]
Another solution would be to enable each filterbank to overwrite the encode/decode. Don't know if this is good software practice.. 
I'm might be ok with the current solution in #2. 
Feedback appreciated.

File: asteroid_filterbanks/enc_dec.py (hunk 1 above)
Comment (popcornell) [reply asteroid_filterbanks/enc_dec.py:152]: Does this work for uneven kernel sizes ( e.g. center=True for free filterbank ) ?

File: asteroid_filterbanks/enc_dec.py (hunk 4 above)
Comment (popcornell) [reply asteroid_filterbanks/enc_dec.py:326]: if it is trimmed we can just suppress the warnings. But if reflect padding is used it will be fine ? Isn't reflect padding as you said used in torch stft ?

File: asteroid_filterbanks/enc_dec.py (hunk 1 above)
Comment (mpariente) [reply asteroid_filterbanks/enc_dec.py:152]: Maybe not, I'd have to check. Also that's why I don't like the `center` everywhere option..

File: asteroid_filterbanks/enc_dec.py (hunk 4 above)
Comment (mpariente) [reply asteroid_filterbanks/enc_dec.py:326]: With this design, you cannot know if it's going to be trimmed, that's the problem.. Because you don't have access to the `length` argument. 
`torch.stft` uses reflect padding, yes. But I don't see why this would solve the problem, sorry

Comment (popcornell) [reply asteroid_filterbanks/enc_dec.py:326]: Forget about the padding. I am not sure I am following. The start seems to be trimmed here right ? So this argument applies only to the tail.

Comment (mpariente) [reply asteroid_filterbanks/enc_dec.py:326]: Yes, true. 
There is not uncertainty on the beginning of the signal: 0 if not center, else kernel_size // 2
But there is uncertainty on the tail, and `torch.istft`'s approach is weird IMO

Comment (popcornell) [reply asteroid_filterbanks/enc_dec.py:326]: Is indeed weird in fact. I don't like it, right now you avoid division by w**2 if too small but that could cause an abrupt discontinuity at the tail no ?

Comment (mpariente) [reply asteroid_filterbanks/enc_dec.py:326]: Yes, completely. 
In torch, they raise an error instead. 
So the with the classic `hanning` window, without center=True, it cannot be `istft`ed :sweat: 

I thought about several cases etc.. And then I realized, if people want to use a "suboptimal" normalization scheme for there stft, they can haha. I won't spend more time than this

Comment (mpariente) [reply asteroid_filterbanks/enc_dec.py:326]: When we raise a warning, either: 
- Torch would have made an error. 
- Torch wouldn't have made an error, and we'll trim it in the next function. 

Anyway, this is only useful for people who already trained a model with torch.stft and want to transfer the model to ONNX or something. So we can assume that the params are good. Otherwise, I think they should use asteroid's stft instead. Much more practical for transfers, for real-time implementation as well.

**Comment by mpariente** [comment]
I'm closing this because most people agree. The center in the encoder and decoder is not useful.

File: asteroid_filterbanks/enc_dec.py (hunk 4 above)
Comment (popcornell) [reply asteroid_filterbanks/enc_dec.py:326]: Agree, also we should write on the docs about this problem so people are aware and can then make their choice.
